### PR TITLE
Add JSX support in type definition

### PR DESCRIPTION
**Pull request by remcohaszing**
This adds TypeScript support to use `xastscript` as JSX. This is just an idea. It’s definitely open for discussion. :)

The interface already supports JSX legacy mode, meaning this was already supported, except for types.

I don’t know if this was intended. Basically any function that accepts a tag name as first argument, props as second argument, and children as spread arguments, can be considered a JSX compatible function. I figured for `xastscript` it makes sense, because it’s for working with XML.

Some quirks:

- JSX elements starting with upper case will be converted to variables, so XML tags starting with lower case can be used directly.
- TypeScript doesn’t support namespaces in JSX.
- Users may request support for the new JSX mode at some point.

### CODE REVIEW

**Comment by wooorm** [comment]
Huh, interesting!

The unist-builder/hastscript/xastscript signatures are indeed designed to be somewhat like `React.createElement`, although they are all based on the earlier hyperscript.

The new JSX transform in React 17 is in my opinion nice for React, but unfortunate for anyone else who handles JSX. For example, it’s React-specific that `children` is a prop (and because they treat it as one, we can never have an actual `children` prop in xast). Another problem, which already exists, is that JSX is coupled to the framework: while both React and Vue support JSX, it’s not possible to write (more complex) JSX that works in both. Anyway, all different things.

---

Code-wise, some of the comment are still missing, or read a bit confusing to me.
This would probably also be useful to describe with an example in the readme?

There should be tests for members in tag names: `<a.b />`, and for attribute expressions: `<a {...b} />`.

Should we somehow handle fragments too? Turn into roots? Only makes sense when it‘s the outer thing, and should be extracted when occurring as a child, and thrown on when having props? 🤔

JSX allows tags/fragments as attribute values too: `<welcome name=<>Venus</> />`, `<welcome name=<span>Pluto</span> />`, which doesn’t make sense for us.

**Comment by wooorm** [comment]
If we’re going to do this, it’d be wise to have actual checks for it in the tests too. But that could be a different PR. And we’d check acorn and others to see how they handle JSX.

**Comment by remcohaszing** [comment]
> Huh, interesting!

My thought exactly! :D

> The unist-builder/hastscript/xastscript signatures are indeed designed to be somewhat like `React.createElement`, although they are all based on the earlier hyperscript.

Only one JSX pragma can be used per file. I think this might be interesting for `hastscript` as well, but not so much for `unist-builder`.

> The new JSX transform in React 17 is in my opinion nice for React, but unfortunate for anyone else who handles JSX. For example, it’s React-specific that `children` is a prop (and because they treat it as one, we can never have an actual `children` prop in xast). Another problem, which already exists, is that JSX is coupled to the framework: while both React and Vue support JSX, it’s not possible to write (more complex) JSX that works in both. Anyway, all different things.

I agree. Let’s forget about this. :)

> Code-wise, some of the comment are still missing, or read a bit confusing to me.

Type definitions for JSX are super confusing. Also JSX type definitions don’t line up with how JSX can be used in React or Preact.

> This would probably also be useful to describe with an example in the readme?

I wanted to check if you think this is a good idea before continueing this. :smiley: 

> There should be tests for members in tag names: `<a.b />`, and for attribute expressions: `<a {...b} />`.

`<a.b />` uses a functional component. This is the same situation as for `<Bar />`.

There is a test for props (`<urlset xmlns={xmlns} />`). I should probably add a test to show only string attributes are allowed. This is defined by `IntrinsicElements`.

> Should we somehow handle fragments too? Turn into roots? Only makes sense when it‘s the outer thing, and should be extracted when occurring as a child, and thrown on when having props? 

See below.

> JSX allows tags/fragments as attribute values too: `<welcome name=<>Venus</> />`, `<welcome name=<span>Pluto</span> />`, which doesn’t make sense for us.

This is disallowed by the `IntrinsicElements` type.

> If we’re going to do this, it’d be wise to have actual checks for it in the tests too. But that could be a different PR.

This would mean the tests need to be processed by Babel or TypeScript. This doesn’t mean it can’t be done.

> And we’d check acorn and others to see how they handle JSX.

Acorn is a JavaScript parser. I don’t see why we should look how they handle JSX. `xastscript` implements the interface, like [react](https://reactjs.org/), [preact](https://preactjs.com/),  or [mini-jsx](https://gitlab.com/appsemble/mini-jsx) (which I wrote, including types).

---

I’ll try to give a brief explanation of what JSX is.

The following 2 blocks of code are identical:

```jsx
/** @jsx x */
import * as x from 'hastscript';

function Component() {}

x('foo', null);
x('foo', { bar: 'baz' });
x('foo', null, x('bar', null), x('baz', null))
x(Component, null);

<foo />
<foo bar="baz" />
<foo>
  <bar />
  <baz />
</foo>
<Component />
```

JSX tag names matching `[a-z_][\da-z_]*` will be converted to string. All other components will be converted to identifiers. Props are simply objects, or `null` if no props are defined.

In TypeScript, the` JSX.Element` type defines the value of the returned JSX node. This is a bug. The return value could be dynamic on the implementation of the JSX function. This isn’t relevant for `xastscript` though.

---

A functional component is a library specific implementation detail. The actual call signature depends on the implementing library. 

In React, a functional component is a function that has the following signature:

```ts
type ReactNode = string | number | boolean | null | JSX.Element | Array<ReactNode>;
function MyComponent(props?: ObjectLikeInterface): ReactNode;
```

In TypeScript, a functional component is a function that has the following signature:

```ts
function MyComponent(props?: ObjectLikeInterface): JSX.Element;
```

This is a bug in TypeScript. The return type of functional components should match the same type as the return value of the JSX implementation, which isn’t true for React and other libraries.

---

A fragment is a functional component that takes no props (or optional props only), whose name is marked as fragment. This means adding support for fragments actually means adding support for functional components.

```jsx
/** @jsxFrag Foo */

// This is a fragment now.
function Foo() {
  return <div />
}

// yields <div></div>
<></>
```

Typically a fragment is used to return wrapped children.

For a fragment implementation `xastscript` the return type would be `Array<xast.Node>`, which is not a valid according to TypeScript, as it doesn’t match `xast.Element`.

---

Now the explanations of the added type definitions:

`Element`: The return type of a JSX expression, as well as the return type of functional components.

`IntrinsicAttributes`: Shared props allowed by all components. I.e. this is used to define `key` in React. This should be set to `xast.Attributes`, but setting this to `never` combined with how `IntrinsicElements` is used, disallows functional and class components while keeping support for string tag names.

`ElementChildrenAttribute`: This interface contains one key. This key defines as what prop children is passed to functional components. This also determines which key in `IntrinsicElements` defines the type of `Children`.

`IntrinsicElements`: is a mapping of known tag name types. Typically this is used to define which HTML or SVG elements support which props. In this case, any tag name may use `xast.Attributes`. The `''` key matches the key in `ElementChildrenAttribute`, meaning it’s used to define the type of JSX children.

**Comment by wooorm** [comment]
> > If we’re going to do this, it’d be wise to have actual checks for it in the tests too. But that could be a different PR.
>
> This would mean the tests need to be processed by Babel or TypeScript. This doesn’t mean it can’t be done.

See GH-4!

**Comment by ChristianMurphy** [comment]
switching to semver patch, since this was added as a feature in #4, this now catches the type definitions up to the implementation

File: types/test.tsx [hunk 1]
```diff
@@ -20,6 +23,41 @@ x('urlset', {xmlns}, x('loc'), x('loc')) // $ExpectType Element
...
+    <loc />
+  </urlset>
+)
+// $ExpectType Element
+const jsx7 = <urlset>{[<loc />, <loc />]}</urlset>
+// $ExpectType Element
+const jsx8 = <urlset>{[]}</urlset>
+
+declare function Bar(props?: Attributes): Element
+const bar = <Bar /> // $ExpectError
```
Comment (ChristianMurphy) [inline types/test.tsx]: It could be good to have both a ts and a tsx test file, to allow the plain ts features to be tested independent of the tsx syntax.
I looked into this in #7, either this PR or that could move forward, I wanted to avoid committing over @remcohaszing in case there is a difference in opinion.

File: types/index.d.ts [hunk 2]
```diff
@@ -21,8 +21,49 @@ declare function xastscript(name: string, ...children: Children[]): Element
...
+   * This defines the prop types for known elements.
+   *
+   * For `xastscript` this defines any string may be used in combination with `xast` `Attributes`.
+   */
+  interface IntrinsicElements {
+    [key: string]: xast.Attributes & {
+      /**
+       * The prop that matches `ElementChildrenAttribute` key defines the type of JSX children, defines the children type.
+       */
+      ''?: Children
```
Comment (ChristianMurphy) [inline types/index.d.ts]: Out of curiosity where did you find documentation on the key `''` part for children attribute?
I scanned https://www.typescriptlang.org/docs/handbook/jsx.html and couldn't find a reference.

File: types/test.tsx [hunk 3]
```diff
@@ -1,3 +1,6 @@
+/** @jsx x */
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import {Attributes, Element} from 'xast'
```
Comment (ChristianMurphy) [inline types/test.tsx]: I don't think these need to be imported, dtslint does some magic to auto detect/import types used in expectations.

File: types/index.d.ts (hunk 2 above)
Comment (remcohaszing) [reply types/index.d.ts]: I figured this out while implementing [mini-jsx](https://gitlab.com/appsemble/mini-jsx/-/blob/master/h.ts#L474-510). Finding out about this was a mix between reading the documentation and reverse engineering `@types/react`.

**Comment by wooorm** [comment]
...soooo, which one should we go with? Both are failing 🤷‍♂️ Anything unresolved things? Other ways I can help?

File: types/test.tsx (hunk 1 above)
Comment (remcohaszing) [reply types/test.tsx]: I split it as you requested. :smile:

File: types/index.d.ts [hunk 4]
```diff
@@ -1,8 +1,8 @@
 // TypeScript Version: 3.7
 
-import {Element, Node, Root} from 'xast'
+import * as xast from 'xast'
 
-type Children = string | Node | number | Children[]
+type Children = string | xast.Node | number | Children[]
```
Comment (remcohaszing) [inline types/index.d.ts:5]: I’m confused about this one. Can any `xast.Node` be a child? I.e. the following is valid:

```ts
h('foo', null, { type: 'whatever' })
```

TBH I don’t even understand why this is exported in `@types/xast`. It’s a direct export of `unist.Node`.

This is unrelated to this change by the way.

Comment (ChristianMurphy) [reply types/index.d.ts:5]: It is valid https://runkit.com/christianmurphy/5fbc319108e0d1001a633d4d

---

> I don’t even understand why this is exported in

"this" as in `Children`?
I see it as a utility type, which is more complex than `Node`.
"this" as in `xast.Node` itself?
It's a convenience type, it doesn't need to be re-exported, but I don't see harm in re-exporting it.
Do you have a particular concern?

File: types/index.d.ts [hunk 5]
```diff
@@ -29,10 +39,16 @@ declare function xastscript(name: null, ...children: Children[]): Root
  */
 declare function xastscript(
   name: string,
-  attributes?: xast.Attributes,
+  attributes?: Attributes,
   ...children: Children[]
 ): xast.Element
 
+/**
+ * This unique symbol is declared to specify they key on which JSX children are passed, without conflicting
```
Comment (wooorm) [inline types/index.d.ts]:
```suggestion
 * This unique symbol is declared to specify the key on which JSX children are passed, without conflicting
```

File: readme.md [hunk 6]
```diff
@@ -231,6 +231,19 @@ var x = require('xastscript')
 console.log(<music />)
 ```
 
+For [TypeScript][], this can be done by setting `"jsx": "react"` and
```
Comment (wooorm) [inline readme.md]: Should it be documented why `jsx` is `react`, even though we have nothing to do with react?

Comment (remcohaszing) [reply readme.md]: I added a link to the TypeScript documentation on configuring JSX. I think this should be sufficient.

File: types/index.d.ts (hunk 4 above)
Comment (remcohaszing) [reply types/index.d.ts:5]: I meant _“this”_ as in `xast.Node` itself.

I moved the issue for fixing `Children` to #8.

I moved the discussion of the removal of `xast.Node` to https://github.com/syntax-tree/unist/discussions/30

File: types/index.d.ts [hunk 7]
```diff
@@ -1,8 +1,8 @@
-// TypeScript Version: 3.7
+// TypeScript Version: 4.0
```
Comment (ChristianMurphy) [inline types/index.d.ts:1]: :thinking: I'm not quite sure how to handle this SemVer wise.
On one hand this should still be compatible with 3.7, your changes are non-breaking.
But on the other hand to use the fragment feature in the tests, this is updated to 4.0, meaning we no longer test previous version.
Thoughts on whether this should be patch or major?

Comment (remcohaszing) [reply types/index.d.ts:1]: Considering this doesn't break functionality for older TypeScript versions, and the fact I haven't had any issues upgrading to TypeScript 4.0 in other projects, I'd say this is semver patch.

The alternative is not to test fragments, which I think is worse.

This could be combined. The types could be added, but the fragment tests could be delayed for a semver major release. Along with the minimal TypeScript version bump.

Comment (wooorm) [reply types/index.d.ts:1]: If the code that runs doesn’t change, the semver version doesn’t change, imo.

I’m fine with:
- fragments in tests, updating the version string there (maybe describing that it’s only for tests), releasing it in a patch
- for now commented out fragments in tests, not updating the version string there, releasing it in a patch

Comment (remcohaszing) [reply types/index.d.ts:1]: IMO this can be released as-is then. :smile:

File: types/index.d.ts [hunk 8]
```diff
@@ -38,6 +38,58 @@ declare function xastscript(
...
   attributes?: Attributes,
   ...children: Children[]
-): Element
+): xast.Element
+
+/**
+ * This unique symbol is declared to specify the key on which JSX children are passed, without conflicting
+ * with the Attributes type.
+ */
+declare const children: unique symbol
```
Comment (ChristianMurphy) [inline types/index.d.ts:47]: huh, that's new https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#unique-symbol
neat!

File: types/index.d.ts [hunk 9]
```diff
@@ -38,6 +38,58 @@ declare function xastscript(
...
+  }
+
+  /**
+   * The key of this interface defines as what prop children are passed.
+   */
+  interface ElementChildrenAttribute {
+    /**
+     * Only the key matters, not the value.
+     */
+    [children]?: any
```
Comment (ChristianMurphy) [inline types/index.d.ts]: if only the key matters, would it be safe to use `unknown` here?

**Review by ChristianMurphy — APPROVED** [review]
Thanks @remcohaszing!

File: types/index.d.ts (hunk 9 above)
Comment (remcohaszing) [reply types/index.d.ts]: I thought I did, but apparently I didn’t :thinking: 

I changed it to `never`, which I think fits even better than `unknown`.

**Review by ChristianMurphy — APPROVED** [review]

**Comment by remcohaszing** [comment]
Ping @wooorm

I think this can be merged and released :smile:

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
thanks, [released](https://github.com/syntax-tree/xastscript/releases/tag/2.1.1)!
